s4e-backend/src/main/java/pl/cyfronet/s4e/controller/InvitationController.java

```diff
@@ -0,0 +1,82 @@
+import pl.cyfronet.s4e.controller.request.CreateInvitationRequest;
+import pl.cyfronet.s4e.controller.response.BasicInstitutionResponse;
+import pl.cyfronet.s4e.event.OnConfirmInvitationEvent;
+import pl.cyfronet.s4e.event.OnRejectInvitationEvent;
+import pl.cyfronet.s4e.event.OnSendInvitationEvent;
+import pl.cyfronet.s4e.service.InvitationService;
+
+import javax.validation.Valid;
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+import static pl.cyfronet.s4e.Constants.API_PREFIX_V1;
+
+@RestController
+@RequestMapping(path = API_PREFIX_V1, produces = APPLICATION_JSON_VALUE)
+@RequiredArgsConstructor
+@Tag(name = "invitation", description = "Invite by email API")
+public class InvitationController {
+    private final InvitationService invitationService;
+    private final ApplicationEventPublisher eventPublisher;
+
+    @Operation(summary = "Send invitation email to user")
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "Invitation was send"),
+            @ApiResponse(responseCode = "400", description = "Incorrect request", content = @Content),
+            @ApiResponse(responseCode = "401", description = "Unauthenticated", content = @Content),
+            @ApiResponse(responseCode = "403", description = "Forbidden", content = @Content),
+            @ApiResponse(responseCode = "404", description = "Not found", content = @Content)
+    })
+    @PostMapping(value = "/institutions/{institution}/invitations", consumes = APPLICATION_JSON_VALUE)
+    public void send(
+            @RequestBody @Valid CreateInvitationRequest request,
+            @PathVariable("institution") String institutionSlug
+    ) throws Exception {
+        invitationService.createInvitationFrom(request, institutionSlug);
+        val invitationEvent = new OnSendInvitationEvent(request.getEmail(), institutionSlug, LocaleContextHolder.getLocale());
+        eventPublisher.publishEvent(invitationEvent);
+    }
+
+    @Operation(summary = "Send invitation confirmation email to user")
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "Invitation have been confirmed"),
+            @ApiResponse(responseCode = "400", description = "Incorrect request", content = @Content),
+            @ApiResponse(responseCode = "401", description = "Unauthenticated", content = @Content),
+            @ApiResponse(responseCode = "404", description = "Not found", content = @Content)
+    })
+    @PostMapping(value = "/invitations/{token}/confirm", consumes = APPLICATION_JSON_VALUE)
+    public BasicInstitutionResponse confirm(@PathVariable("token") String token) {
+        invitationService.confirmInvitation(token);
+        val confirmInvitationEvent = new OnConfirmInvitationEvent(token, LocaleContextHolder.getLocale());
+        eventPublisher.publishEvent(confirmInvitationEvent);
+
+        return invitationService.findInstitutionBy(token, BasicInstitutionResponse.class);
```

Comment:
Just to be on the safe side, I'd move this invitationService call before ::publishEvent. The invitiation may be already deleted, based on the threads scheduling.